src/iptraf.8
.TH IPTRAF 8 "IPTraf Help Page"
.SH NAME
iptraf \- Interactive Colorful IP LAN Monitor
.SH SYNOPSIS
.BR iptraf " { [ " \-f " ] [ " \-q " ] [ " \-u " ] [ { " \-i 
.IR iface " | " 
.BR \-g " | " \-d
.IR iface " | "
.BR \-s
.IR iface " | "
.BR \-z
.IR iface " | "
.BR \-l 
.IR iface " } [ " 
.BR \-t
.IR timeout " ] [ "
.BR \-B " [ "
.BR \-L
.IR logfile " ] ] ] | [ "
.BR \-h " ] }"
.br
.SH DESCRIPTION
.B iptraf
is an ncurses-based IP LAN monitor that generates various network statistics including TCP info, UDP counts, ICMP and OSPF information, Ethernet load info, node stats, IP checksum errors, and others.
.PP
If the 
.Biptraf 
command is issued without any command-line options, the program comes up in interactive mode, with the various facilities accessed through the main menu.

.SH OPTIONS
These options can also be supplied to the command:
.TP
.BI "\-i " iface
immediately start the IP traffic monitor on the specified interface, or
all interfaces if "\-i all" is specified
.TP
.B "\-g"
immediately start the general interface statistics
.TP
.BI "\-d " iface
allows you to immediately start the detailed on the indicated interface (iface)
.TP
.BI "\-s " iface
allows you to immediately monitor TCP and UDP traffic on the specified interface (iface)
.TP
.BI "\-z " iface
shows packet counts by size on the specified interface
.TP
.BI "\-l " iface
start the LAN station monitor on the specified interface, or all LAN
interfaces if "\-l all" is specified
.TP
.BI "\-t " timeout
tells IPTraf to run the specified facility for only
.I timeout
minutes.  This option is used only with one of the above parameters.
.TP
.B "\-B"
redirect standard output to /dev/null, closes standard input, and forks
the program into the background.  Can be used only with one of the
facility invocation parameters above.  Send the backgrounded process a
USR2 signal to terminate.
.TP
.B "\-L logfile"
allows you to specify an alternate log file name.  The default log file
name is based on either the interface selected (detailed interface
statistics, TCP/UDP service statistics, packet size breakdown), or the
instance of the facility (IP traffic monitor, LAN station monitor).  If a
path is not specified, the log file is placed in
.B /var/log/iptraf
.TP
.B "\-f"
clears all locks and counters, causing this instance of IPTraf to think
it's the first one running.  This should only be used to recover from
an abnormal termination or system crash.
.TP
.B "\-u"
allow use of unsupported interfaces as ethernet devices.  This is needed if
you changed the name of an interface (ex: ip link set eth0 name foo0)
.TP
.BI "\-q"
no longer needed, maintained only for compatibility.
.TP
.B "\-h"
shows a command summary
.SH SIGNALS

 SIGUSR1 - rotates log files while program is running
 SIGUSR2 - terminates an IPTraf process running in the background.

.SH FILES
 /var/log/iptraf/*.log - log file
 /var/lib/iptraf/* - important IPTraf data files

.SH SEE ALSO
 Documentation/* - complete documentation written by the author
.br

.SH AUTHOR
Gerard Paul Java (riker@mozcom.com)

.SH MANUAL AUTHOR
Frederic Peters (fpeters@debian.org), using iptraf \-h
General manual page modifications by Gerard Paul Java (riker@mozcom.com)



CHANGES
Fork IPTraf-ng
==============================
* Thu Mar 04 2010 Nikola Pajkovsky <npajkovs@redhat.com> - 1.0.1
- remove dereferencing type-punned pointer will break strict-aliasing rules
- convmacaddr(): fix assignment casting. One more warning gone.
- ipmon(), servmon(): unnecessary casting is gone. Fixes strict-aliasing.
    
    gcc will no longer warn about type-punned pointers that break
    strict-aliasing
- Fix compilation issues and crash within LAN station monitor
    reorder includes to successfully compile on older kernels
    
    embrace changes from Debian patch iptraf_3.0.0-6.diff:
    - convmacaddr(): fix string to numeric format conversion
	(fixes the crash)
    - remove superfluous casting
------------------------------------------------------------------
* Mon Mar 01 2010 Nikola Pajkovsky <npajkovs@redhat.com> - 1.0.0
- compile with -std=gnu99 -pedantic -Wall -Werror
- add support for devices(used patches from gentoo, debian and fedora):

- add support for devices(used patches from gentoo, debian and fedora):
        vlan, hsi, ctc, ath, bond, ra, bnep, dsl, modem, ni, br, tap, dummy,  vmnet
  add longer intefaces names(18 chars)
  add iptraf -u - show unsupported interfaces(fpeters@debian.org)
  check if macro is defined ICMP6_DST_UNREACH_NOTNEIGHBOR(jer@gentoo.org)

- add ipv6 support
- remove wierd Setup compilation and replaced it by autotools
- import original iptraf-3.0.0
==============================
==============================
CHANGES File for IPTraf 3.0.0

Changes to IPTraf 2.7.0 and new features in IPTraf 3.0.0

	New filter behavior.  Except for TCP traffic in the IP traffic
	monitor, filters now do not automatically match reverse packets
	for TCP and UDP IPTraf-wide.  Rather, each filter entry has
	a field which tells IPTraf whether to match packets flowing
	in the direction opposite that specified.

	The filters for non-TCP, non-UDP IP traffic (ICMP, IGRP, OSPF,
	etc.) which never automatically matched packets flowing in the
	opposite direction, now have that same option field.  This way
	related packets (like ICMP echo request/echo reply) can be
	matched with a single entry.

	Because reverse-matching is no longer the default IPTraf-wide,
	the labels are now changed to read Source and Destination.

        Default value for blank address filter fields is now 0.0.0.0,
	rather than 255.255.255.255. Fields are therefore no longer
	pre-filled with 0.0.0.0.

	Miscellaneous IP filter entries feature a field for other IP
	protocols not specifically indicated in the dialog.  The user
	must enter a comma-separated list of individual protocols or a 
	range.  IP protocols are defined in the /etc/protocols file.

	The IP traffic monitor consults the /etc/protocols file for 
	miscellaneous IP packets for the protocol names.  Previously
	recognized protocols (ICMP, UDP, OSPF, etc) are still looked up 
	internally for performance reasons.

        The filter rule selection now indicates the mask in CIDR format
	(e.g. 10.1.0.0/16) for clarity and to save screen space.

        Filter selection list box is now alphabetically sorted.

	Likewise, the CIDR notation can be used when entering IP address
	data.  However the CIDR notation is translated into a mask and
	discarded.  Subsequent editing of the filter will show the 
	corresponding mask.

	Changed color coding for unknown IP packets (those looked up
	from /etc/services to bright white on blue (instead of yellow on
	red, which looked like "errors").

	Added internal recognition for L2TP, IPSec AH, and IPSec ESP
	packets.

        Changed size of the IP traffic monitor's TCP hash table to 1033
	buckets.  Prime number used to improve hash efficiency.  

	A new function tx_box() has been added to the screen support
	library as a solution to the ncurses box() function not accepting
	the color set by wattrset(), at least on Red Hat 7.3.  All calls
	to box() have been replaced with this tx_box() instead.  It takes
	exactly the same parameters.

	Added support for tun and brg (tunneling and bridging) interfaces.
	Thanks to Marcio Gomes <tecnica_at_microlink.com.br>.

	Modified logging options.  The -L parameter now works with any
	command-line invocation of a facility, even in foreground mode.

	Added -I command-line parameter to override logging interval
	configuration option.

	(Thanks to the contributors of the -I and -L patches. I lost your
 	emails when SEUL reinstalled. Please acknowledge. Thanks.
	
	Corrected promiscuous mode control code.  It ignored Token Ring 
	interfaces.

Changes to IPTraf 2.6.1 and new features in IPTraf 2.7.0

	Corrected bug wherein the detailed interface statistics
	did not filter out the packets based on the selected
	interface.  Thanks to the members of the mailing list for
	this.

	Corrected minor interface name comparison bugs in the
	general interface statistics and TCP/UDP service statistics.

	Corrected stale locks when IPTraf did not start due to an
	improper terminal size.

	Added support for additional DVB interfaces sm2*, sm3*, penta*.

	Added support for wireless LAN interfaces (wlan*, wvlan*).

	Fixed segfault that occurs when /proc/net/dev is empty or

src/iptraf.c
/*
===========================================================================
IPTraf
An IP Network Statistics Utility
Written by Gerard Paul Java <riker@seul.org>
Copyright (c) Gerard Paul Java 1997-2004

Version 3.0.0
Main Module

---------------------------------------------------------------------------
This software is open-source; you may redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed WITHOUT ANY WARRANTY; without even the
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License in the included COPYING file for
details.
---------------------------------------------------------------------------
*/

#define MAIN_MODULE

#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <curses.h>
#include <panel.h>
#include <signal.h>
#include <string.h>
#include <locale.h>
#include <time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <dirent.h>
#include <errno.h>
#include <menurt.h>
#include <winops.h>
#include <msgboxes.h>
#include "dirs.h"
#include "deskman.h"
#include "fltdefs.h"
#include "fltselect.h"
#include "fltmgr.h"
#include "fltedit.h"
#include "ifstats.h"
#include "serv.h"
#include "options.h"
#include "promisc.h"
#include "externs.h"
#include "instances.h"
#include "tcptable.h"
#include "log.h"
#include "attrs.h"
#include "rvnamed.h"
#include "logvars.h"
#include "getpath.h"


#include "config.h"

#define WITHALL 1
#define WITHOUTALL 0

const char *ALLSPEC = "all";

/*
 * Important globals used throughout the
 * program.
 */
int exitloop = 0;
int daemonized = 0;
int facility_running = 0;
int is_first_instance;
char active_facility_lockfile[64];
char active_facility_countfile[64];
int accept_unsupported_interfaces = 0;
char graphing_filter[80];

extern void about();

void press_enter_to_continue(void)
{
    fprintf(stderr, "Press Enter to continue.\n");
    getchar();
}

void clearfiles(char *prefix, char *directory)
{
    DIR *dir;
    struct dirent *dir_entry;
    char target_name[80];

    dir = opendir(directory);

    if (dir == NULL) {
        fprintf(stderr, "\nUnable to read directory %s\n%s\n",
                directory, strerror(errno));
        press_enter_to_continue();
        return;
    }

    do {
        dir_entry = readdir(dir);
        if (dir_entry != NULL) {
            if (strncmp(dir_entry->d_name, prefix, strlen(prefix)) == 0) {
                snprintf(target_name, 80, "%s/%s", directory,
                         dir_entry->d_name);
                unlink(target_name);
            }
        }
    } while (dir_entry != NULL);

    closedir(dir);
}

void removetags(void)
{
    clearfiles("iptraf", LOCKDIR);
}

void remove_sockets(void)
{
    clearfiles(SOCKET_PREFIX, WORKDIR);
}

/*
 * Handlers for the TERM signal and HUP signals.  There's nothing we can do
 * for the KILL.
 */

void term_signal_handler(int signo)
{
    erase();
    refresh();
    endwin();

    if (signo != SIGHUP)
        fprintf(stderr, "IPTraf process %u exiting on signal %d\n\n",
                getpid(), signo);

    if (active_facility_lockfile[0] != '\0') {
        unlink(active_facility_lockfile);
        adjust_instance_count(PROCCOUNTFILE, -1);
        if (active_facility_countfile[0] != '\0')
            adjust_instance_count(active_facility_countfile, -1);
    }

    if (is_first_instance)
        unlink(IPTIDFILE);

    exit(1);
}

/* 
 * Handler for the SIGSEGV, Segmentation Fault.  Tries to clear the screen
 * and issue a better message than "Segmentation fault".  May not always
 * clean up properly.
 */

void segvhandler()
{
    erase();
    refresh();
    endwin();
    fprintf(stderr, "Fatal: memory allocation error\n\n");
    fprintf(stderr,
            "If you suspect a bug, please report the exact circumstances under which this\n");
    fprintf(stderr,
            "error was generated.  If possible, include gdb or strace data which may point\n");
    fprintf(stderr,
            "out where the error occured.  Bug reports may be sent in to iptraf@seul.org.\n\n");
    fprintf(stderr,
            "An attempt will be made to clear all lock files, but if stale lock files\n");
    fprintf(stderr,
            "remain, exit all other instances of IPTraf and restart with the -f\n");
    fprintf(stderr, "command-line parameter.\n\n");
    fprintf(stderr, "IPTraf process %u aborting on signal 11.\n\n",
            getpid());

    if (active_facility_lockfile[0] != '\0')
        unlink(active_facility_lockfile);

    if (is_first_instance)
        unlink(IPTIDFILE);

    if (active_facility_lockfile[0] != '\0') {
        unlink(active_facility_lockfile);
        adjust_instance_count(PROCCOUNTFILE, -1);
        if (active_facility_countfile[0] != '\0')
            adjust_instance_count(active_facility_countfile, -1);
    }

    exit(2);
}

/*
 * USR2 handler.  Used to normally exit a daemonized facility.
 */

void term_usr2_handler()
{
    exitloop = 1;
}

void init_break_menu(struct MENU *break_menu)
{
    tx_initmenu(break_menu, 6, 20, (LINES - 6) / 2, COLS / 2,
                BOXATTR, STDATTR, HIGHATTR, BARSTDATTR, BARHIGHATTR,
                DESCATTR);
    tx_additem(break_menu, " By packet ^s^ize",
               "Displays packet counts by packet size range");
    tx_additem(break_menu, " By TCP/UDP ^p^ort",
               "Displays packet and byte counts by service port");
    tx_additem(break_menu, NULL, NULL);
    tx_additem(break_menu, " E^x^it menu", "Return to main menu");
}

/*
 * Get the ball rolling: The program interface routine.
 */

void program_interface(struct OPTIONS *options,
                       int opt, char *optarg, int facilitytime)
{
    struct MENU menu;
    struct MENU break_menu;

    int endloop = 0;
    int row = 1;
    int break_row = 1;
    int aborted;
    int break_aborted;

    struct filterstate ofilter;
    struct ffnode *fltfiles;

    char ifname[18];
    char *ifptr = NULL;
    struct porttab *ports;

    draw_desktop();

    attrset(STATUSBARATTR);
    mvprintw(0, 1, "IPTraf");

    /*
     * Load saved filter or graphing filter if specified
     */
    if (graphing_logfile[0] != '\0') {
        loadfilterlist(&fltfiles);
        memset(&ofilter, 0, sizeof(struct filterstate));
        loadfilter(pickfilterbyname(fltfiles, graphing_filter),
                   &(ofilter.fl), FLT_RESOLVE);
    } else {
        loadfilters(&ofilter);
        indicate("");
    }

    loadaddports(&ports);

    if (opt == 0) {
        attrset(STATUSBARATTR);
        mvprintw(LINES - 1, 1, PLATFORM);
        about();

        tx_initmenu(&menu, 13, 35, (LINES - 14) / 2, (COLS - 35) / 2,
                    BOXATTR, STDATTR, HIGHATTR, BARSTDATTR, BARHIGHATTR,
                    DESCATTR);

        tx_additem(&menu, " IP traffic ^m^onitor",
                   "Displays current IP traffic information");
        tx_additem(&menu, " General interface ^s^tatistics",
                   "Displays some statistics for attached interfaces");
        tx_additem(&menu, " ^D^etailed interface statistics",
                   "Displays more statistics for a selected interface");
        tx_additem(&menu, " Statistical ^b^reakdowns...",
                   "Facilities for traffic counts by packet size or TCP/UDP port");
        tx_additem(&menu, " ^L^AN station monitor",
                   "Displays statistics on detected LAN stations");
        tx_additem(&menu, NULL, NULL);
        tx_additem(&menu, " ^F^ilters...",
                   "Allows you to select traffic display and logging criteria");
        tx_additem(&menu, NULL, NULL);
        tx_additem(&menu, " C^o^nfigure...",
                   "Set various program options");
        tx_additem(&menu, NULL, NULL);
        tx_additem(&menu, " E^x^it", "Exits program");

        endloop = 0;

        do {
            tx_showmenu(&menu);
            tx_operatemenu(&menu, &row, &aborted);

            switch (row) {
            case 1:
                selectiface(ifname, WITHALL, &aborted);
                if (!aborted) {
                    if (strcmp(ifname, "") != 0)
                        ifptr = ifname;
                    else
                        ifptr = NULL;

                    ipmon(options, &ofilter, 0, ifptr);
                }
                break;
            case 2:
                ifstats(options, &ofilter, 0);
                break;
            case 3:
                selectiface(ifname, WITHOUTALL, &aborted);
                if (!aborted)
                    detstats(ifname, options, 0, &ofilter);
                break;
            case 4:
                break_row = 1;
                init_break_menu(&break_menu);
                tx_showmenu(&break_menu);
                tx_operatemenu(&break_menu, &break_row, &break_aborted);

                switch (break_row) {
                case 1:
                    selectiface(ifname, WITHOUTALL, &aborted);
                    if (!aborted)
                        packet_size_breakdown(options, ifname, 0,
                                              &ofilter);
                    break;
                case 2:
                    selectiface(ifname, WITHOUTALL, &aborted);
                    if (!aborted)
                        servmon(ifname, ports, options, 0, &ofilter);
                    break;
                case 4:
                    break;
                }
                tx_destroymenu(&break_menu);
                break;
            case 5:
                selectiface(ifname, WITHALL, &aborted);
                if (!aborted) {
                    if (strcmp(ifname, "") != 0)
                        ifptr = ifname;
                    else
                        ifptr = NULL;
                    hostmon(options, 0, ifptr, &ofilter);
                }
                break;
            case 7:
                config_filters(&ofilter);
                savefilters(&ofilter);
                break;
            case 9:
                setoptions(options, &ports);
                saveoptions(options);
                break;
            case 11:
                endloop = 1;
                break;
            }
        } while (!endloop);

        tx_destroymenu(&menu);
    } else {
        switch (opt) {
        case 'i':
            if ((strcmp(optarg, ALLSPEC) == 0)
                || (strcmp(optarg, "") == 0))
                ifptr = NULL;
            else
                ifptr = optarg;

            ipmon(options, &ofilter, facilitytime, ifptr);
            break;
        case 'g':
            ifstats(options, &ofilter, facilitytime);
            break;
        case 'd':
            detstats(optarg, options, facilitytime, &ofilter);
            break;
        case 's':
            servmon(optarg, ports, options, facilitytime, &ofilter);
            break;
        case 'z':
            packet_size_breakdown(options, optarg, facilitytime, &ofilter);
            break;
        case 'l':
            if ((strcmp(optarg, ALLSPEC) == 0)
                || (strcmp(optarg, "") == 0))
                ifptr = NULL;
            else
                ifptr = optarg;

            hostmon(options, facilitytime, ifptr, &ofilter);
            break;
        }
    }

    destroyporttab(ports);
    erase();
    update_panels();
    doupdate();
}


/*
 * Command-line help facility.
 */

void commandhelp()
{
    printf("\nSyntax:\n");
    printf
        ("    iptraf [ -f ] [ -u ] [ { -i iface | -g | -d iface | -s iface | -z iface |\n");
    printf
        ("           -l iface } [ -t timeout ] [ -B ] [ -L logfile ] [-I interval] ] \n\n");
    printf
        ("Issue the iptraf command with no parameters for menu-driven operation.\n");
    printf("These options can also be supplied to the command:\n\n");
    printf
        ("-i iface    - start the IP traffic monitor (use \"-i all\" for all interfaces)\n");
    printf("-g          - start the general interface statistics\n");
    printf
        ("-d iface    - start the detailed statistics facility on an interface\n");
    printf
        ("-s iface    - start the TCP and UDP monitor on an interface\n");
    printf("-z iface    - shows the packet size counts on an interface\n");
    printf
        ("-l iface    - start the LAN station monitor (\"-l all\" for all LAN interfaces)\n");
    printf
        ("-B          - run in background (use only with one of the above parameters)\n");
    printf
        ("-t timeout  - when used with one of the above parameters, tells\n");
    printf
        ("              the facility to run only for the specified number of\n");
    printf("              minutes (timeout)\n");
    printf
        ("-L logfile  - specifies an alternate log file for any direct invocation\n");
    printf
        ("              of a facility from the command line.  The log is placed in\n");
    printf("              %s if path is not specified.\n", LOGDIR);
    printf
        ("-I interval - specifies the log interval for all facilities except the IP\n");
    printf("              traffic monitor.  Value is in minutes.\n");
    printf
        ("-f          - clear all locks and counters.  Use with great caution.\n");
    printf
        ("              Normally used to recover from an abnormal termination.\n\n");
    printf
        ("-u          - allow use of unsupported interfaces as ethernet devices.\n");
    printf("IPTraf %s Copyright (c) Gerard Paul Java 1997-2004\n",
           VERSION);
}

int first_instance()
{
    int fd;

    fd = open(IPTIDFILE, O_RDONLY);

    if (fd < 0)
        return !0;
    else {
        close(fd);
        return 0;
    }
}

void mark_first_instance()
{
    int fd;

    fd = open(IPTIDFILE, O_CREAT | O_WRONLY | O_TRUNC, S_IRUSR | S_IWUSR);
    if (fd < 0) {
        fprintf(stderr, "\nWarning: unable to tag this process\r\n");
        press_enter_to_continue();
        return;
    }
    close(fd);
}

/*
 * Main routine
 */

int main(int argc, char **argv)
{
    struct OPTIONS options;
    int opt = 0;
    int command = 0;
    char keyparm[12];
    int facilitytime = 0;
    int current_log_interval = 0;

#ifndef ALLOWUSERS
    if (getuid() != 0) {
        fprintf(stderr, "\nIPTraf Version %s\n", VERSION);
        fprintf(stderr, "Copyright (c) Gerard Paul Java 1997-2004l\n\n");
        fprintf(stderr,
                "This program can be run only by the system administrator\n\n");
        exit(1);
    }
#endif

    setlocale(LC_ALL, "");

    strcpy(current_logfile, "");
    strcpy(graphing_logfile, "");
    strcpy(graphing_filter, "");

    /*
     * Parse command line
     */

    if (argc > 1) {
        do {
            opterr = 0;
            opt = getopt(argc, argv, "i:gd:s:z:l:hfqt:BL:uI:G:F:");

            if (opt == 'h') {
                commandhelp();
                exit(0);
            } else if (opt == 'f') {
                removetags();
                remove_sockets();
            } else if (opt == 't') {
                facilitytime = atoi(optarg);
                if (facilitytime == 0) {
                    fprintf(stderr, "\nInvalid time value\n\n");
                    exit(1);
                }
            } else if (opt == 'B') {
                daemonized = 1;
                setenv("TERM", "linux", 1);
            } else if (opt == 'L') {
                if (strchr(optarg, '/') != NULL)
                    strncpy(current_logfile, optarg, 80);
                else
                    strncpy(current_logfile, get_path(T_LOGDIR, optarg),
                            80);
            } else if (opt == 'q') {
                /* -q parameter now ignored, maintained for compatibility */
            } else if (opt == 'u') {
                accept_unsupported_interfaces = 1;
            } else if (opt == 'I') {
                current_log_interval = atoi(optarg);
                if (current_log_interval == 0)
                    fprintf(stderr, "Invalid log interval value\n");

                exit(1);
            } else if (opt == 'G') {
                if (strchr(optarg, '/') != NULL)
                    strncpy(graphing_logfile, optarg, 80);
                else
                    strncpy(graphing_logfile, get_path(T_LOGDIR, optarg),
                            80);

                daemonized = 1;
            } else if (opt == 'F') {
                strncpy(graphing_filter, optarg, 80);
            } else if (opt == '?') {
                fprintf(stderr,
                        "\nInvalid option or missing parameter, use iptraf -h for help\n\n");
                exit(1);
            } else if (opt != -1) {
                if (optarg != 0) {
                    bzero(keyparm, 12);
                    strncpy(keyparm, optarg, 11);
                } else
                    strcpy(keyparm, "");

                command = opt;
            }
        } while ((opt != '?') && (opt != -1));
    }
    is_first_instance = first_instance();

    if ((getenv("TERM") == NULL) && (!daemonized)) {
        fprintf(stderr, "Your TERM variable is not set.\n");
        fprintf(stderr, "Please set it to an appropriate value.\n");
        exit(1);
    }

    if (graphing_logfile[0] != '\0' && graphing_filter[0] == '\0') {
        fprintf(stderr, "Specify an IP filter name with -F\n");
        exit(1);
    }

    loadoptions(&options);

    /*
     * If a facility is directly invoked from the command line, check for
     * a daemonization request
     */

    if ((daemonized) && (command != 0)) {
        switch (fork()) {
        case 0:                /* child */
            setsid();
            FILE* fd = NULL;
            fd = freopen("/dev/null", "w", stdout);  /* redirect std output */
            fd = freopen("/dev/null", "r", stdin);   /* redirect std input */
            fd = freopen("/dev/null", "w", stderr);  /* redirect std error */
            signal(SIGUSR2,  term_usr2_handler);

            if (graphing_logfile[0] != '\0')
                options.logging = 0;    /* if raw logging is specified */
            else                /* then standard logging is disabled */
                options.logging = 1;
            break;
        case -1:               /* error */
            fprintf(stderr,
                    "\nFork error, IPTraf cannot run in background\n\n");
            exit(1);
        default:               /* parent */
            exit(0);
        }
    }
#ifdef SIMDAEMON
    daemonized = 1;
    freopen("/dev/null", "w", stdout);  /* redirect std output */
    freopen("/dev/null", "r", stdin);
    freopen("/dev/null", "w", stderr);
#endif

    /* Check whether LOCKDIR exists (/var/run is on a tmpfs in Ubuntu) */
    if(access(LOCKDIR,F_OK) != 0)
    {
	    if(mkdir(LOCKDIR, 0700) == -1)
        {
	        fprintf(stderr, "Cannot create %s: %s", LOCKDIR, strerror(errno));
	        exit(1);
        }

	    if(chown(LOCKDIR, 0, 0) == -1) {
	        fprintf(stderr, "Cannot change owner of %s: %s", LOCKDIR, strerror(errno));
	        exit(1);
	    }
    }

    initscr();

    if ((LINES < 24) || (COLS < 80)) {
        endwin();
        fprintf(stderr,
                "\nThis program requires a screen size of at least 80 columns by 24 lines\n");
        fprintf(stderr, "Please resize your window\n\n");
        exit(1);
    }

    mark_first_instance();

    signal(SIGTERM, term_signal_handler);
    signal(SIGHUP, term_signal_handler);
    signal(SIGSEGV, segvhandler);
    signal(SIGTSTP, SIG_IGN);
    signal(SIGINT, SIG_IGN);
    signal(SIGUSR1, SIG_IGN);

    start_color();
    standardcolors(options.color);
    noecho();
    nonl();
    cbreak();

#ifndef DEBUG
    curs_set(0);
#endif

    /*
     * Set logfilename variable to NULL if -L was specified without an
     * appropriate facility on the command line.
     */

    if (command == 0)
        strcpy(current_logfile, "");

    /*
     * If by this time the logfile is still acceptable, obtain the
     * logspan from the command line if so specified.
     */

    if (current_logfile[0] != '\0') {
        options.logging = 1;
        if (current_log_interval != 0) {
            options.logspan = current_log_interval;
        }
    }

    program_interface(&options, command, keyparm, facilitytime);

    endwin();

    if (is_first_instance)
        unlink(IPTIDFILE);

    return (0);
}
